Added RegisterUserPrecheck function

NJavokhir opened the pull request:
## Test Cases Checklist

### SECTION 1: RESOURCE SERVER (http://localhost:5001/)
- [x] Created a new user on the Layer8 Portal

### SECTION 2: WE'VE GOT POEMS (http://localhost:5173/)
- [x] Login as 'tester', '1234'
- [ ] Choose login with layer8 
- [ ] Logging in with the newly registered credentials from Section 1 succeeds in popup
- [ ] User chooses to share their new "Username" & "Country" from the Layer8 Resource Server
- [ ] Choose poems 1 - 3 works correctly 

### SECTION 3: IMSHARER (http://localhost:5174/)
- [ ] Uploads an image works
- [ ] Reload leads to instant reload (demonstrating proper caching)

interestIngc commented on server/resource_server/service/service.go:
```diff
@@ -264,3 +265,18 @@ func (s *service) UpdateUserPassword(username string, newPassword string, salt s
 	return s.repository.UpdateUserPassword(username, hashedPassword)
 }
+
+func (s *service) RegisterUserPrecheck(req dto.RegisterUserPrecheckDTO, iterCount int) (models.RegisterUserPrecheckResponseOutput, error) {
```
It would be better to make this function return just an error, since the goal of the function is to register a user, the output should be either a success or a failure value, which we can express in the ``error`` type. Adding the ``RegisterUserPrecheckResponseOutput`` generation breaks the Single Responsibility principle (https://en.wikipedia.org/wiki/Single-responsibility_principle) since this method would be able to get modified for two reasons: 
1. changing business logic to register a user;
3. changing format of the response.

Thus, it's better to extract logic related to generating ``RegisterUserPrecheckResponseOutput`` into controller.go and make this handler just return an error type.

NJavokhir replied:
@interestIngc  Random salt is generating here on service function. If we return only error type , how can we respond random salt to frontend?

interestIngc commented on server/resource_server/controller/controller.go:
```diff
@@ -554,3 +555,44 @@ func validateHttpMethod(w http.ResponseWriter, actualMethod string, expectedMeth
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+
+	if err := json.NewEncoder(w).Encode(registerUserPrecheckResp); err != nil {
```
Here we can generate the response model (i.e. ``models.RegisterUserPrecheckResponseOutput``) and then call the ``utils.BuildResponse`` method to populate response with a status header and data (salt and iterationCount). An example for a similar type of response is on line 465 in ``GetUsageStats`` method.

NJavokhir replied:
I did it

interestIngc commented on server/resource_server/models/user.go:
```diff
@@ -11,6 +11,7 @@ type User struct {
 	VerificationCode string `gorm:"column:verification_code; not null" json:"verification_code"`
 	ZkKeyPairId      uint   `gorm:"column:zk_key_pair_id; not null" json:"zk_key_pair_id"`
 	PublicKey        []byte `gorm:"column:public_key; not null" json:"public_key"`
+	IterationCount   int    `gorm:"column:iterationCount;" json:"iterationCount"`
```
Better to put column name in camel case, i.e. ``iteration_count``

interestIngc commented on server/resource_server/repository/repository.go:
```diff
@@ -324,3 +324,17 @@ func (r *Repository) IsBackendURIExists(backendURL string) (bool, error) {
 	return count > 0, nil
 }
+
+func (r *Repository) RegisterPrecheckUser(req dto.RegisterUserPrecheckDTO, salt string, iterCount int) (string, int, error) {
```
No need to return salt and iterCount from this method, since these values are already known to the caller (since they are passed here as an input). We can just make this method return an ``error`` type which would reflect whether or not user's data was persisted successfully.

interestIngc commented on sp_mocks/wgp/backend/package-lock.json:
```diff
@@ -1054,8 +1054,7 @@
       "resolved": "https://registry.npmjs.org/layer8_middleware/-/layer8_middleware-0.0.22.tgz",
-      "integrity": "sha512-HFvLDM8LNe/ijeiUzjclWGYUNqTUEPMKO9ua1jeXfdgx0NWFe+2WBv6FLuLeAUl7UO/Z43Um/zodPIegKsuZmA==",
-      "license": "GPL-2.0"
+      "integrity": "sha512-HFvLDM8LNe/ijeiUzjclWGYUNqTUEPMKO9ua1jeXfdgx0NWFe+2WBv6FLuLeAUl7UO/Z43Um/zodPIegKsuZmA=="
```
Do we need changes in this file for the current PR?

NJavokhir replied:
No, we don't need it.

interestIngc commented on sp_mocks/wgp/frontend/package-lock.json:
```diff
@@ -1235,8 +1235,7 @@
       "resolved": "https://registry.npmjs.org/layer8_interceptor/-/layer8_interceptor-0.0.22.tgz",
-      "integrity": "sha512-67RYEm1IwuQ56cJuzm6uDA6RFBJeaLDxPOEB3WQVVLOxtxRlyhpYdWZx57Q/S/3QWnBhvbB+kMakF5Ni37vmjw==",
-      "license": "GPL-2.0"
+      "integrity": "sha512-67RYEm1IwuQ56cJuzm6uDA6RFBJeaLDxPOEB3WQVVLOxtxRlyhpYdWZx57Q/S/3QWnBhvbB+kMakF5Ni37vmjw=="
```
Same here

interestIngc reviewed (changes requested):
I've left the initial review. Great work thus far!

interestIngc commented:
Thanks Javokhir! Can you please revert merging veronika/scram-base branch and just rebase it on development?

NJavokhir commented:
> Thanks Javokhir! Can you please revert merging veronika/scram-base branch and just rebase it on development?

Done! Reverted and rebased. Now, I am solving the conflics.

interestIngc commented:
> > Thanks Javokhir! Can you please revert merging veronika/scram-base branch and just rebase it on development?
> 
> Done! Reverted and rebased. Now, I am solving the conflicts.

Thanks! But there are still Huzaifa's commits on this branch for some reason, can you please remove them from this branch? Let me know if you need help with this.

huzaifamk commented:
How are my commits appearing in this branch, I haven't pushed nor worked under this branch :eyes:

interestIngc commented on test/cypress/package-lock.json:
```diff
@@ -1955,4 +1955,4 @@
     }
   }
-}
+}
```
We can revert this change as it's not related to the PR (looks like it just adds an empty line)

interestIngc reviewed (commented):
Almost there! Please address these comments and then PR is ready to be merged!

NJavokhir commented on server/resource_server/service/service.go:
```diff
@@ -2,14 +2,15 @@ package service
 	"globe-and-citizen/layer8/server/resource_server/utils"
 	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
```
yes, for validating signature

interestIngc commented on server/resource_server/controller/controller.go:
```diff
@@ -554,3 +558,47 @@ func validateHttpMethod(w http.ResponseWriter, actualMethod string, expectedMeth
+
+	iterCount, err := strconv.Atoi(os.Getenv("SCRAM_ITERATION_COUNT"))
+	if err != nil {
+		utils.HandleError(w, http.StatusBadRequest, "Invalid iteration count configuration", err)
```
In this case we should return http.StatusInternalServerError (500 status code)

NJavokhir replied:
Done!

interestIngc commented on server/resource_server/controller/controller_test.go:
```diff
@@ -1852,3 +1857,144 @@ func setMockServiceInContext(req *http.Request) *http.Request {
+	req := httptest.NewRequest("POST", "/api/register-user-precheck", bytes.NewBuffer(reqBody))
+	req = req.WithContext(context.WithValue(req.Context(), "service", mockService))
+	req.Header.Set("Content-Type", "application/json")
+
```
Here you should set `t.Setenv("SCRAM_ITERATION_COUNT", "4096")`, currently the test fails because of that (I think it tries to parse an integer from an empty string). 
In case of missing json fields, the returned status code should be 400 (bad request).

interestIngc reviewed (approved):
Great work! The PR is ready to be merged!
